src/tests/dotnet-monitor/EndpointInfoSourceTests.cs

```diff
@@ -0,0 +1,242 @@
+
+            await using (var execution1 = StartTraceeProcess("LoggerRemoteTest", transportName))
+            {
+                await newEndpointInfoHelper.WaitForNewEndpointInfoAsync(TimeSpan.FromSeconds(5));
+
+                execution1.Start();
+
+                endpointInfos = await GetEndpointInfoAsync(source);
+
+                var endpointInfo = Assert.Single(endpointInfos);
+                VerifyConnection(execution1.TestRunner, endpointInfo);
+
+                _outputHelper.WriteLine("Stopping tracee.");
+            }
+
+            await Task.Delay(TimeSpan.FromSeconds(1));
+
+            endpointInfos = await GetEndpointInfoAsync(source);
+
+            Assert.Empty(endpointInfos);
+        }
+
+        private TestServerEndpointInfoSource CreateServerSource(out string transportName)
+        {
+            transportName = ReversedServerHelper.CreateServerTransportName();
+            _outputHelper.WriteLine("Starting server endpoint info source at '" + transportName + "'.");
+            return new TestServerEndpointInfoSource(transportName, _outputHelper);
+        }
+
+        private RemoteTestExecution StartTraceeProcess(string loggerCategory, string transportName = null)
+        {
+            _outputHelper.WriteLine("Starting tracee.");
+            string exePath = CommonHelper.GetTraceePath("EventPipeTracee", targetFramework: "net5.0");
+            return RemoteTestExecution.StartProcess(exePath + " " + loggerCategory, _outputHelper, transportName);
+        }
+
+        private async Task<IEnumerable<IEndpointInfo>> GetEndpointInfoAsync(ServerEndpointInfoSource source)
+        {
+            _outputHelper.WriteLine("Getting endpoint infos.");
+            using CancellationTokenSource cancellationSource = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+            return await source.GetEndpointInfoAsync(cancellationSource.Token);
+        }
+
+        /// <summary>
+        /// Verifies basic information on the connection and that it matches the target process from the runner.
+        /// </summary>
+        private static void VerifyConnection(TestRunner runner, IEndpointInfo endpointInfo)
+        {
+            Assert.NotNull(runner);
+            Assert.NotNull(endpointInfo);
+            Assert.Equal(runner.Pid, endpointInfo.ProcessId);
+            Assert.NotEqual(Guid.Empty, endpointInfo.RuntimeInstanceCookie);
+            Assert.NotNull(endpointInfo.Endpoint);
+        }
+
+        /// <summary>
+        /// Helper class for waiting for notification of a new connection from the connection source.
+        /// This aids in allowing to wait for a new connection with a timeout rather than waiting
+        /// for a specified amount of time and then testing for the new connection (and possibly repeating
+        /// if a new connection was not found).
+        /// </summary>
+        private sealed class NewEndpointInfoHelper : IDisposable
```

Comment:
The entire NewEndpointInfoHelper can be folded into TestServerEndpointInfoSource.

Reply:
Fixed